app/modules/sale.js

```diff
@@ -1,58 +1,199 @@
 // @flow
 import { wallet, api } from 'neon-js'
+import { flatten } from 'lodash'
 
-import { showErrorNotification, showInfoNotification, showSuccessNotification } from './notifications'
-import { getWIF, LOGOUT } from './account'
+import {
+  showErrorNotification,
+  showInfoNotification,
+  hideNotification
+} from './notifications'
+import {
+  getWIF,
+  LOGOUT,
+  getAddress,
+  getIsHardwareLogin,
+  getSigningFunction,
+  getPublicKey
+} from './account'
 import { getNetwork } from './metadata'
-import { getNEO } from './wallet'
-import { toBigNumber, toNumber } from '../core/math'
+import { getNEO, getGAS } from './wallet'
+
+import { toNumber } from '../core/math'
 import asyncWrap from '../core/asyncHelper'
+import { ASSETS } from '../core/constants'
+import {
+  validateMintTokensInputs,
+  validateOldMintTokensInputs
+} from '../core/sale'
+import { oldMintTokens } from '../core/oldMintTokens'
 
-// TODO: Rewrite this function
+const MESSAGES = {
+  PARTICIPATION_FAILED: 'Sale participation failed, please check your script hash again.'
+}
 
-export const participateInSale = (neoToSend: string, scriptHash: string) => async (dispatch: DispatchType, getState: GetStateType) => {
+export const participateInSale = (
+  neoToSend: string,
+  gasToSend: string,
+  scriptHash: string,
+  gasCost: string = '0'
+) => async (dispatch: DispatchType, getState: GetStateType) => {
   const state = getState()
   const wif = getWIF(state)
-  const NEO = getNEO(state)
+  const publicKey = getPublicKey(state)
+  const NEO = toNumber(getNEO(state))
+  const GAS = toNumber(getGAS(state))
   const net = getNetwork(state)
+  const address = getAddress(state)
+  const isHardwareLogin = getIsHardwareLogin(state)
+  const signingFunction = getSigningFunction(state)
 
   const account = new wallet.Account(wif)
-  if (!toBigNumber(neoToSend).isInteger()) {
-    dispatch(showErrorNotification({ message: 'You cannot send fractional NEO to a token sale.' }))
+  const neoToMint = toNumber(neoToSend)
+  const gasToMint = toNumber(gasToSend)
```

Comment:
I think we should keep these as a string when calling `validateMintTokensInputs`, then let `toBigNumber` be called internally from there.